src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -695,11 +695,18 @@ void c_typecheck_baset::typecheck_expr_builtin_offsetof(exprt &expr)
       // still need to typecheck index
       typecheck_expr(index);
 
-      exprt sub_size=size_of_expr(type.subtype(), *this);
+      auto sub_size_opt = size_of_expr(type.subtype(), *this);
+
+      if(!sub_size_opt.has_value())
+      {
+        error().source_location = expr.source_location();
+        error() << "offsetof failed to determine array element size" << eom;
+        throw 0;
```

Comment:
Do not throw unstructured exceptions

Reply:
Fair point, but the C and C++ front-ends do this all over the place at the moment. It should be fixed, but it's outside the scope of this PR.